File: Chains/GrabAndThrow.cpp
#include "GrabAndThrow.h"

void GrabAndThrow::PressButtons()
{
    //If we're dashing, then jump cancel the grab
    if(m_state->m_memory->player_two_action == DASHING ||
        m_state->m_memory->player_two_action == RUNNING)
     {
         m_controller->pressButton(Controller::BUTTON_Y);
         return;
     }

    //Do the grab if we're in a state to do so and haven't yet
    if(m_state->m_memory->player_two_action != GRAB &&
        m_state->m_memory->player_two_action != GRAB_RUNNING &&
        m_state->m_memory->player_two_action != THROW_FORWARD &&
        m_state->m_memory->player_two_action != THROW_BACK &&
        m_state->m_memory->player_two_action != THROW_UP &&
        m_state->m_memory->player_two_action != THROW_DOWN &&
        m_state->m_memory->player_two_action != GRAB_PULLING &&
        m_state->m_memory->player_two_action != GRAB_WAIT &&
        m_grabbedYet == false)
    {
        m_grabbedYet = true;
        m_controller->releaseButton(Controller::BUTTON_Y);
        m_controller->pressButton(Controller::BUTTON_Z);
        return;
    }

    if(m_state->m_memory->player_two_action == GRAB_PULLING)
    {
        m_controller->emptyInput();
        return;
    }

    //Pummel our opponent for some free damage
    if(!m_pummeledYet && m_state->m_memory->player_two_action == GRAB_WAIT)
    {
        m_controller->pressButton(Controller::BUTTON_A);
        m_pummeledYet = true;
        return;
    }

    //Perform a throw
    if(m_state->m_memory->player_two_action == GRAB_WAIT)
    {
        m_controller->releaseButton(Controller::BUTTON_A);
        switch(m_direction)
        {
            case UP_THROW:
            {
                m_controller->tiltAnalog(Controller::BUTTON_MAIN, .5, 1);
                break;
            }
            case DOWN_THROW:
            {
                m_controller->tiltAnalog(Controller::BUTTON_MAIN, .5, 0);
                break;
            }
            case BACK_THROW:
            {
                m_controller->tiltAnalog(Controller::BUTTON_MAIN, m_state->m_memory->player_two_facing ? 0 : 1, .5);
                break;
            }
            case FORWARD_THROW:
            {
                m_controller->tiltAnalog(Controller::BUTTON_MAIN, m_state->m_memory->player_two_facing ? 1 : 0, .5);
                break;
            }
        }
        return;
    }

    //If we get the throw, then let go for a frame
    if(m_state->m_memory->player_two_action == THROW_UP ||
        m_state->m_memory->player_two_action == THROW_DOWN ||
        m_state->m_memory->player_two_action == THROW_BACK ||
        m_state->m_memory->player_two_action == THROW_FORWARD)
    {
        m_controller->emptyInput();
        return;
    }

    //If we miss the grab, let go
    if(m_state->m_memory->player_two_action_frame > 15)
    {
        m_controller->emptyInput();
        return;
    }
}

bool GrabAndThrow::IsInterruptible()
{
    if((m_state->m_memory->player_two_action == GRAB ||
        m_state->m_memory->player_two_action == GRAB_RUNNING) &&
        m_state->m_memory->player_two_action_frame > 20)
    {
        return true;
    }

    if(m_state->m_memory->player_two_action == THROW_UP ||
        m_state->m_memory->player_two_action == THROW_DOWN ||
        m_state->m_memory->player_two_action == THROW_BACK ||
        m_state->m_memory->player_two_action == THROW_FORWARD)
    {
        return true;
    }

    //Safety return
    uint frame = m_state->m_memory->frame - m_startingFrame;
    if(frame >= 120)
    {
        return true;
    }
    return false;}

GrabAndThrow::GrabAndThrow(THROW_DIRECTION direction)
{
    m_direction = direction;
    m_grabbedYet = false;
    m_pummeledYet = false;
    if(m_state->m_memory->player_one_percent < 20)
    {
        m_pummeledYet = true;
    }

}

GrabAndThrow::~GrabAndThrow()
{
}

File: Chains/grabandthrow.py
import melee
from melee.enums import Action, Button
from Chains.chain import Chain
from enum import Enum

class THROW_DIRECTION(Enum):
    UP = 0
    DOWN = 1
    FORWARD = 2
    BACK = 3

# Grab and throw opponent
class GrabAndThrow(Chain):
    def __init__(self, direction=THROW_DIRECTION.DOWN):
        self.direction = direction

    def step(self, gamestate, smashbot_state, opponent_state):
        controller = self.controller

        self.interruptible = False

        # If we already pressed Z last frame, let go
        if controller.prev.button[Button.BUTTON_Z]:
            controller.empty_input()
            return

        if smashbot_state.action == Action.GRAB and smashbot_state.action_frame > 12:
            controller.empty_input()
            self.interruptible = True
            return

        if smashbot_state.action == Action.LANDING_SPECIAL:
            controller.empty_input()
            self.interruptible = True
            return

        # If we need to jump cancel, do it
        jcstates = [Action.DOWN_B_GROUND_START, Action.DOWN_B_GROUND]
        if (smashbot_state.action in jcstates) and smashbot_state.action_frame >= 3:
            controller.press_button(Button.BUTTON_Y)
            controller.release_button(Button.BUTTON_Z)
            return

        # Grab on knee bend
        if smashbot_state.action == Action.KNEE_BEND:
            # Let go of Z if we already had it pressed
            if controller.prev.button[Button.BUTTON_Z]:
                controller.release_button(Button.BUTTON_Z)
                return
            controller.press_button(Button.BUTTON_Z)
            return

        # Do the throw
        if smashbot_state.action in [Action.GRAB_WAIT, Action.GRAB_PULLING]:
            if self.direction == THROW_DIRECTION.DOWN:
                controller.tilt_analog(Button.BUTTON_MAIN, .5, 0)
            if self.direction == THROW_DIRECTION.UP:
                controller.tilt_analog(Button.BUTTON_MAIN, .5, 1)
            if self.direction == THROW_DIRECTION.FORWARD:
                controller.tilt_analog(Button.BUTTON_MAIN, int(smashbot_state.facing), .5)
            if self.direction == THROW_DIRECTION.BACK:
                controller.tilt_analog(Button.BUTTON_MAIN, int(not smashbot_state.facing), .5)
            self.interruptible = True
            return

        # Do the grab
        # Let go of Z if we already had it pressed
        if controller.prev.button[Button.BUTTON_Z]:
            controller.release_button(Button.BUTTON_Z)
            return
        controller.press_button(Button.BUTTON_Z)
